modules/javafx.web/src/main/java/com/sun/webkit/network/PublicSuffixes.java

```diff
@@ -78,94 +84,116 @@ static boolean isPublicSuffix(String domain) {
-        }
+
+        Rules rules = Rules.getRules(domain);
+        return rules == null ? false : rules.match(domain);
     }
 
-    /**
-     * Loads the public suffix list from a given resource.
-     */
-    private static Map<String,Rule> loadRules(String resourceName) {
-        logger.finest("resourceName: [{0}]", resourceName);
-        Map<String,Rule> result = null;
-
-        InputStream is = PublicSuffixes.class.getResourceAsStream(resourceName);
-        if (is != null) {
-            BufferedReader reader = null;
-            try {
-                reader = new BufferedReader(new InputStreamReader(is, "UTF-8"));
-                result = loadRules(reader);
-            } catch (IOException ex) {
-                logger.warning("Unexpected error", ex);
-            } finally {
+    private static class Rules {
+
+        private final Map<String, Rule> rules = new HashMap<>();
+
+        private Rules(InputStream is) throws IOException {
+            InputStreamReader isr = new InputStreamReader(is, "UTF-8");
+            BufferedReader reader = new BufferedReader(isr);
+
+            String line;
+            int type = reader.read();
+            while (type != -1 && (line = reader.readLine()) != null) {
+                Rule rule;
+                if (line.startsWith("!")) {
+                    line = line.substring(1);
+                    rule = Rule.EXCEPTION_RULE;
+                } else if (line.startsWith("*.")) {
+                    line = line.substring(2);
+                    rule = Rule.WILDCARD_RULE;
+                } else {
+                    rule = Rule.SIMPLE_RULE;
+                }
                 try {
-                    if (reader != null) {
-                        reader.close();
-                    }
-                } catch (IOException ex) {
-                    logger.warning("Unexpected error", ex);
+                    line = IDN.toASCII(line, IDN.ALLOW_UNASSIGNED);
+                } catch (Exception ex) {
+                    logger.warning(String.format("Error parsing rule: [%s]", line), ex);
+                    continue;
                 }
+                rules.put(line, rule);
+                type = reader.read();
+            }
+            if (logger.isLoggable(Level.FINEST)) {
+                logger.finest("rules: {0}", toLogString(rules));
             }
-        } else {
-            logger.warning("Resource not found: [{0}]",
-                    resourceName);
         }
 
-        result = result != null
-                ? Collections.unmodifiableMap(result)
-                : Collections.<String,Rule>emptyMap();
-        if (logger.isLoggable(Level.FINEST)) {
-            logger.finest("result: {0}", toLogString(result));
+        static Rules getRules(String domain) {
+            String tld = getTopLevelDomain(domain);
+            if (tld.isEmpty()) {
+                return null;
+            }
+            return rulesCache.computeIfAbsent(tld, k -> createRules(tld));
```

Comment:
If the public_suffixes file cannot be loaded (for example, if it is removed from the JDK), then this causes a warning to be logged on every call to this (basically on every load of any web site). It might be better to call `createRules` from a static init block.

Reply:
If `createRules` is called from a static init block, then all the public suffix rules (for all TLDs) must be loaded during initialization. Another approach would be to set a `fileNotFound` flag which could be set to true in `getPubSuffixStream` and call `Rules.getRules` only when the value is `false`.